[None][doc] add VisualGen MGMN NVL72 tech blog (#15420)
NVIDIA/TensorRT-LLM: docs/source/blogs/tech_blog/blog25_Scaling_Video_Generation_Across_NVL72_Rack_with_TensorRT-LLM.md

The VAE is handled separately from the denoiser axes. Instead of being carved from the denoiser mesh, its group is built by **overlaying a fresh device group on top of the existing world** (the `cfg × tp × cp × ulysses` mesh), which lets VAE parallelism scale independently of the DiT (see [Scaling the VAE Decoder](#scaling-the-vae-decoder) above). Its one constraint is that the VAE ranks fit within a single CFG group, or span the whole world: after the guidance combine the latent handed to the decoder lives on a single CFG branch, so a VAE group straddling both branches would mix the conditional and unconditional copies and its gather/halo collectives would be incoherent.

For the full property catalog (every `*_group`, `*_rank`, the flattened `seq_mesh`, and the `to_llm_mapping()` bridge that lets Linear layers reuse the LLM TP path), see [`mapping.py`](https://github.com/NVIDIA/TensorRT-LLM/blob/163be837f3e82d092eb33704747a93fb419e7099/tensorrt_llm/_torch/visual_gen/mapping.py).

## NVL72 Scaling Results

> All performance numbers in this section use the same measured workload and publish one distributed recipe: `examples/visual_gen/configs/wan22_t2v_bf16_gb200_nvl72.yml`.

Reviewer: I suppose you’re planning to upload this file later, since it doesn’t seem to exist at the moment. Correct?

Author: Thanks for catching! Uploaded

Reviewer 2: Better to point to the GitHub config URL (I understand that there are some blogs are referencing content assuming that we are in TRTLLM_REPO_ROOT).

Author: how do we use github url if config is only added in this PR? The URL will link to this branch right?

Reviewer 2: Something like https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/visual_gen/configs/cosmos3-super-4gpu.yaml

I think it should be fine that the link is invalid before PR merging.

Author: Updated

The full recipe behind these numbers is shown below:

```yaml
attention_config:
  backend: FA4
parallel_config:
  cfg_size: 2            # CFG: conditional / unconditional streams on disjoint halves
  ulysses_size: 4        # Ulysses head sharding
  attn2d_size: [3, 3]    # Attention2D context parallelism (row × col)
  parallel_vae_size: 8   # spatial VAE sharding (saturates at ~8)
compilation_config:
  resolutions:
    - [720, 1280]        # warmup at the benchmarked shape (request size 1280x720)
  num_frames:
    - 80                 # 5-second clip at 16 fps
```

### Cosmos3-Super

> The same runtime scales a very different model unchanged. Cosmos3-Super is a 64B Mixture-of-Transformers with 64 attention heads, generating a longer 189-frame clip — so each denoising step is an even larger dense prefill than Wan. Only the workload and the per-width recipe change; the CFG, Ulysses, Attention2D, and parallel-VAE knobs are identical (`examples/visual_gen/configs/cosmos3_t2v_bf16_gb200_nvl72.yml`)

Reviewer: This _nvl72.yml as well~

Author: Thanks, added

| Configuration | GPUs | LPIPS↓ | ΔLPIPS vs floor | PSNR↑ | SSIM↑ | Cosine↑ |
| :--- | ---: | ---: | ---: | ---: | ---: | ---: |
| _floor_ (2nd single-GPU run) | 1 | 0.121 | — | 25.18 | 0.878 | 0.981 |
| `ulysses2` | 2 | 0.187 | +0.066 | 24.34 | 0.796 | 0.974 |
| `cfg2` | 2 | 0.200 | +0.079 | 23.83 | 0.777 | 0.970 |
| `cfg2·ulysses2` | 4 | 0.196 | +0.075 | 23.72 | 0.783 | 0.970 |
| `ulysses4` | 4 | 0.198 | +0.077 | 23.87 | 0.777 | 0.969 |
| `attn2d 2×1·ulysses2` | 4 | 0.198 | +0.077 | 24.00 | 0.777 | 0.970 |
| `attn2d 1×2·ulysses2` | 4 | 0.209 | +0.088 | 23.45 | 0.768 | 0.968 |
| `attn2d 2×2` | 4 | 0.220 | +0.099 | 23.36 | 0.759 | 0.967 |

> To drill into that nondeterminism — and to confirm each parallel configuration is measured against the right baseline — we first isolate it on a single GPU. Wan runs in BF16 with two performance knobs, FA4 attention and `torch.compile`, both of which use nondeterministic reductions, so the same prompt and seed need not produce identical pixels and the iterative denoiser compounds those tiny differences across its steps. To size that effect, we hold the model and seed fixed and change one implementation detail at a time:

Reviewer 3: please mention that `torch.use_deterministic_algorithms(True)` can help remove this non-zero LPIPS noise floor and produce identical outputs.

Author: Thanks, will do. Does `torch.use_deterministic_algorithms(True)` decrease perf?

# Scaling Video Generation Across NVL72 rack with TensorRT-LLM

By NVIDIA TensorRT LLM Team

## Abstract

> Modern Diffusion Transformer (DiT) video generation models such as Wan 2.2 T2V-A14B and Cosmos3-Super are computationally expensive: generating a single 5-second 720p clip takes over five minutes. Almost all of that time is the DiT denoising loop, which runs tens of full forward passes, each over a 70-150k-token sequence. Unlike LLM decode, every denoising step in DiT is a "prefill" - classifier-free guidance (CFG) evaluates the transformer twice, and the VAE decoder has its own peak both in compute and in activation memory. At these sequence lengths, the dense bidirectional attention — quadratic in the sequence — dominates the per-step compute cost. Pipeline parallelism cannot amortize a single-step single-request workload. Tensor parallelism reduces per-layer compute effectively within a node but adds communication in every layer including the FFN. Sequence-parallel methods restrict communication to inside the attention operation and directly attack the dominant cost, which is why the sequence is the natural multi-node scaling axis.

Reviewer 2: ```suggestion
Modern Diffusion Transformer (DiT) video generation models such as [Wan 2.2 T2V-A14B](https://huggingface.co/Wan-AI/Wan2.2-T2V-A14B-Diffusers) and [Cosmos3-Super](https://huggingface.co/nvidia/Cosmos3-Super) model) are computationally expensive: generating a single 5-second 720p clip takes over five minutes. Almost all of that time is the DiT denoising loop, which runs tens of full forward passes, each over a 70-150k-token sequence. Unlike LLM decode, every denoising step in DiT is a "prefill" - classifier-free guidance (CFG) evaluates the transformer twice, and the VAE decoder has its own peak both in compute and in activation memory. At these sequence lengths, the dense bidirectional attention — quadratic in the sequence — dominates the per-step compute cost. Pipeline parallelism cannot amortize a single-step single-request workload. Tensor parallelism reduces per-layer compute effectively within a node but adds communication in every layer including the FFN. Sequence-parallel methods restrict communication to inside the attention operation and directly attack the dominant cost, which is why the sequence is the natural multi-node scaling axis.
```

> This post describes how the `VisualGen` runtime in TensorRT-LLM scales the DiT denoising loop and the VAE decoder from a single NVIDIA B200 to a full GB200 NVL72 rack using techniques like \- **CFG**, **Ulysses parallelism**, **Ring Attention**, **Attention2D context parallelism**, and **parallel VAE** \- unified under a single PyTorch `DeviceMesh`. The first four shard the DiT's CFG streams and token sequence, while parallel VAE scales the decoder independently. All axes are declarative knobs on `ParallelConfig`; no model code changes are required to go from 1 GPU to 72\.

Reviewer 2: ```suggestion
This post describes how the `VisualGen` runtime in TensorRT-LLM scales the DiT denoising loop and the VAE decoder from a single NVIDIA B200 to a full GB200 NVL72 rack using techniques like - **CFG**, **Ulysses parallelism**, **Ring Attention**, **Attention2D context parallelism**, and **parallel VAE** - unified under a single PyTorch `DeviceMesh`. The first four shard the DiT's CFG streams and token sequence, while parallel VAE scales the decoder independently. All axes are declarative knobs on `ParallelConfig`; no model code changes are required to go from 1 GPU to 72.
```

If the `\` symbols are intended? Not sure if they are generated by the gdoc exporting. Maybe we can ask agent to cleanup the unintended escapes in a markdown.

Author: removed thanks for catching

> On Wan 2.2 T2V-A14B (a 5-second 1280×720 video, 40 steps), the `CFG=2 × Ulysses=4 × Attention2D 3×3` recipe shrinks the DiT denoising loop by **\~53×** going from a single NVIDIA B200 to a full GB200 NVL72 rack \- which in turn drives a **\~41× end-to-end** speedup. End-to-end trails the denoise speedup because of a fixed tail that stays constant as the DiT is sharded wider, setting the floor by Amdahl's law.

Reviewer 2: ```suggestion
On Wan 2.2 T2V-A14B (a 5-second 1280×720 video, 40 steps), the $CFG=2 \times Ulysses=4 \times Attention2D 3\times3$ recipe shrinks the DiT denoising loop by **\~53x** going from a single NVIDIA B200 to a full GB200 NVL72 rack \- which in turn drives a **\~41× end-to-end** speedup. End-to-end trails the denoise speedup because of a fixed tail that stays constant as the DiT is sharded wider, setting the floor by Amdahl's law.
```

Not a big deal, maybe also consider to use tex for the formulas?

Author: done

Pipeline parallelism, on the other hand, is a poor fit for a different reason. PP hides its inter-stage communication by microbatch pipelining, but that only pays off with many requests in flight. Video generation here is latency-bound on a single request at a time, so at any moment all but one pipeline stage would sit idle. Hence, PP buys no speedup on the metric that matters for these workloads.

This post focuses on the sequence and CFG axes as the primary multi-node scaling story; TP composes with all of them and is out of scope here due to its relatively poor efficiency across multiple nodes. 

> Finally, the VAE decoder at the end of the loop has its own activation-memory peak that is *larger* than any single DiT step's activation footprint, and lives outside the transformer entirely.

Reviewer 2: What about quick intro about how Parallel VAE works?

Author: yep, added

Sweeping the best recipe at every width gives the same near-linear denoise-loop scaling as Wan:

<p align="center">
  <img src="../media/tech_blog25_cosmos3_scaling_lines.png" alt="Cosmos3-Super DiT denoise-loop latency vs ideal linear scaling on GB200 NVL72, with per-point parallel efficiency" width="1080">
</p>

<p align="center"><sub><em>Figure 8. Cosmos3-Super DiT denoise-loop scaling on GB200 NVL72 (best config at each width, GB200 1-GPU baseline). The loop tracks ideal linear scaling closely \- ~49× at 72 GPUs \- because CFG and sequence parallelism shard the transformer almost perfectly.</em></sub></p>

As with Wan, the denoise loop scales almost linearly under the *same* recipe, which proves that the sequence-length sharding pattern for the DiT generalizes across models — even at a very different model size and sequence length. End-to-end efficiency falls off faster because it folds in everything outside the DiT, including the longer clip's larger mp4 encode and transport/persist; those fixed per-request costs become the dominant share once the denoise loop collapses to ~7s (Amdahl's law).

## Serving with trtllm-serve

> The same `--extra_visual_gen_options` YAML drives `trtllm-serve` for online serving. Distributed multi-node serving is handled by the `benchmark_visual_gen_mgmn_distributed.sh` script under `examples/visual_gen/serve/`:

Reviewer 2: ```suggestion
The same `--visual_gen_args` YAML drives `trtllm-serve` for online serving. Distributed multi-node serving is handled by the `benchmark_visual_gen_mgmn_distributed.sh` script under `examples/visual_gen/serve/`:
```

I don't see `benchmark_visual_gen_mgmn_distributed.sh` in the PR. If we plan to ship such as a separate script, better to name like `benchmark_visual_gen_gb200_nvl72.sh`

Author: the benchmark script is here - https://github.com/NVIDIA/TensorRT-LLM/blob/main/examples/visual_gen/serve/benchmark_visual_gen_mgmn_distributed.sh
its not exactly what we used for the benchmarks since that was from automation but this is an example of how to benchmark visual gen server running on multi-gpus

```shell
export CONTAINER_IMAGE=/path/to/tensorrt-llm.sqsh
export PROJECT_ROOT=/path/to/TensorRT-LLM
export MODEL=Wan-AI/Wan2.2-T2V-A14B-Diffusers

> export SERVER_CONFIG=examples/visual_gen/configs/wan22_nvl72.yml

Reviewer 2: update to a default path?

Author: what is the default path?

Reviewer 2: Something like `examples/visual_gen/configs/cosmos3_t2v_bf16_gb200_nvl72.yaml` since we have renamed the YAML path

Author: Ah I see. Yes updated

The result is near-ideal scaling of the DiT denoising steps. On Wan 2.2 T2V-A14B, the `CFG=2 × Ulysses=4 × Attention2D 3×3` recipe shrinks the DiT denoise loop ~53× from a single B200 to a full GB200 NVL72 rack (~41× end-to-end), and the *same* recipe applies unchanged to the 64B Cosmos3-Super (~49× denoise), confirming that the sequence-length sharding pattern generalizes across very different model sizes and sequence lengths. Concretely, a video clip that once took over five minutes to generate now finishes in ~9 seconds on the full GB200 NVL72 rack — and we expect the same approach to carry over to the next generation of longer, higher-resolution video models.

> ## Picking the Right Configuration

Reviewer 2: ```suggestion
## Picking the Performant Configuration
```
